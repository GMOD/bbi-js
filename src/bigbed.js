const BBI = require('./bbi')

export default class BigBed extends BBI {
  getFeatures(refName, start, end, opts = {}) {
    const chrName = this.renameRefSeq(refName)
    let view

    view = this.getView(1)
    if (!view) {
      return null
    }

    return view.readWigData(chrName, start, end)
  }

  parseBigBedBlock(bytes, startOffset) {
    const data = this.window.bwg.newDataView(bytes, startOffset)

    let offset = 0
    while (offset < bytes.byteLength) {
      const chromId = data.getUint32(offset)
      const start = data.getInt32(offset + 4)
      const end = data.getInt32(offset + 8)
      offset += 12
      if (chromId !== this.chr) {
        console.warn('BigBed block is out of current range')
        return
      }

      let rest = ''
      while (offset < bytes.byteLength) {
        const ch = data.getUint8(offset)
        offset += 1
        if (ch !== 0) {
          rest += String.fromCharCode(ch)
        } else {
          break
        }
      }

      const featureData = this.parseBedText(start, end, rest)
      featureData.id = `bb-${startOffset + offset}`
      this.maybeCreateFeature(start, end, featureData)
    }
  }


}
