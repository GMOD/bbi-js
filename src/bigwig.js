const BBI = require('./bbi')

export default class BigWig extends BBI {
  getFeatures(refName, start, end, opts = {}) {
    const chrName = this.renameRefSeq(refName)
    let view

    if (opts.basesPerSpan) {
      view = this.getView(1 / opts.basesPerSpan)
    } else if (opts.scale) {
      view = this.getView(opts.scale)
    } else {
      view = this.getView(1)
    }

    if (!view) {
      return null
    }

    return view.readWigData(chrName, start, end)
  }
}
