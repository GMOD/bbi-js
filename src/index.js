import BBI from './bbi'
import BigWig from './bigwig'
import BigBed from './bigbed'

export { BBI, BigWig, BigBed }
