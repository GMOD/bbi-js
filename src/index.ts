export { BigWig } from './bigwig'
export { BigBed } from './bigbed'
export { Feature, Header, RequestOptions } from './bbi'
