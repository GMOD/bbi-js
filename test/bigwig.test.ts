/* eslint @typescript-eslint/explicit-function-return-type: 0 */
import { BigWig, Header } from '../src/'

class HalfAbortController {
  public signal: any

  public constructor() {
    this.signal = { aborted: false }
  }

  public abort(): void {
    this.signal.aborted = true
  }
}

interface ExtendedHeader extends Header {
  iWasMemoized: boolean
}

describe('bigwig formats', () => {
  it('loads bedgraph bigwig file', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox.bw'),
    })
    const feats1 = await ti.getFeatures('ctgA', 0, 100, { scale: 1 })
    const feats2 = await ti.getFeatures('ctgA', 0, 100, { scale: 0.01 })
    const feats3 = await ti.getFeatures('ctgA', 0, 100, { scale: 0.001 })
    const feats4 = await ti.getFeatures('ctgA', 2000, 2100, { scale: 0.001 })
    expect(feats1).toMatchSnapshot()
    expect(feats2).toMatchSnapshot()
    expect(feats3).toMatchSnapshot()
    expect(feats4).toMatchSnapshot()
  })
  it('loads variable step bigwig', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/variable_step.bw'),
    })
    const feats1 = await ti.getFeatures('chr1', 0, 51, { scale: 1 })
    const feats2 = await ti.getFeatures('chr1', 0, 52, { scale: 1 })
    expect(feats1).toMatchSnapshot()
    expect(feats1[feats1.length - 1].start).toEqual(47)
    expect(feats2[feats2.length - 1].start).toEqual(51)
  })
  it('loads simple fixedstep bigwig', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox_microarray.bw'),
    })
    const feats1 = await ti.getFeatures('ctgA', 0, 1000, { scale: 1 })
    const feats2 = await ti.getFeatures('ctgA', 0, 1000, { scale: 0.01 })
    const feats3 = await ti.getFeatures('ctgA', 0, 1000, { scale: 0.001 })
    const feats4 = await ti.getFeatures('ctgA', 2000, 2100, { scale: 0.001 })
    expect(feats1).toMatchSnapshot()
    expect(feats2).toMatchSnapshot()
    expect(feats3).toMatchSnapshot()
    expect(feats4).toMatchSnapshot()
  })
  it('loads fixedstep bigwig with starts', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/fixedStep.bw'),
    })
    const feats1 = await ti.getFeatures('chr1', 0, 1000, { scale: 1 })
    const feats2 = await ti.getFeatures('chr1', 11000, 12000, { scale: 1 })
    expect(feats1).toMatchSnapshot()
    expect(feats2).toMatchSnapshot()
  })
  it('inside file deeply', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox.bw'),
    })
    const feats5 = await ti.getFeatures('ctgA', 20000, 21000)
    expect(feats5.slice(10, 20)).toMatchSnapshot()
  })

  it('missing data', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox.bw'),
    })
    const feats = await ti.getFeatures('ctgA', 4200, 5600)
    expect(feats.length).toEqual(1401)
    expect(feats.slice(10, 20)).toMatchSnapshot()
    expect(feats.slice(1000, 1010)).toMatchSnapshot()
  })

  it('loads a larger bigwig file at different scales', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/cow.bw'),
    })
    const feats1 = await ti.getFeatures('GK000001.2', 2000000, 2100000, {
      scale: 1,
    })
    const feats2 = await ti.getFeatures('GK000001.2', 2000000, 2100000, {
      scale: 0.01,
    })
    const feats3 = await ti.getFeatures('GK000001.2', 2000000, 2100000, {
      scale: 0.001,
    })
    const feats4 = await ti.getFeatures('GK000001.2', 2000000, 2100000, {
      scale: 0.001,
    })
    const feats5 = await ti.getFeatures('GK000001.2', 2000000, 2100000, {
      scale: 0.00001,
    })
    const f4max = Math.max(...feats4.map(s => s.score))
    const f5max = Math.max(...feats5.map(s => s.maxScore || 0))
    expect(f4max).toEqual(f5max)
    expect(feats5).toMatchSnapshot() // summary block
    expect(feats1.slice(10, 20)).toMatchSnapshot()
    expect(feats2.slice(10, 20)).toMatchSnapshot()
    expect(feats3.slice(10, 20)).toMatchSnapshot()
    expect(feats4.slice(10, 20)).toMatchSnapshot()
  })

  it('performs regularization', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox.bw'),
      renameRefSeqs: ref => ref.replace('contig', 'ctg'),
    })
    const feats = await ti.getFeatures('contigA', 4200, 5600)
    expect(feats.length).toEqual(1401)
  })

  it('matches bigWigToBedGraph', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/cow.bw'),
    })
    const feats = await ti.getFeatures('GK000001.2', 1000000, 1001000)

    // match bigWigToBedGraph output approximately, the start and ends are not clipped like bedGraphToBigWig does though
    const ret = `GK000001.2\t999003\t1000189\t0
    GK000001.2\t1000189\t1000210\t1
    GK000001.2\t1000210\t1000245\t2
    GK000001.2\t1000245\t1000276\t1
    GK000001.2\t1000276\t1000716\t0
    GK000001.2\t1000716\t1000803\t2
    GK000001.2\t1000803\t1000933\t0
    GK000001.2\t1000933\t1000972\t2
    GK000001.2\t1000972\t1000977\t1
    GK000001.2\t1000977\t1001000\t0`
      .split('\n')
      .map(s => {
        const r = s.split('\t')
        return {
          start: +r[1],
          end: +r[2],
          score: +r[3],
        }
      })
    expect(feats).toEqual(ret)
  })

  it('test memoize', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/cow.bw'),
    })
    const ret = (await ti.getHeader()) as ExtendedHeader
    ret.iWasMemoized = true
    const ret2 = (await ti.getHeader()) as ExtendedHeader
    expect(ret2.iWasMemoized).toEqual(true)
  })

  it('abort loading a bigwig file', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox.bw'),
    })
    const aborter = new HalfAbortController()
    const indexDataP = ti.getHeader({ signal: aborter.signal })
    aborter.abort()
    await expect(indexDataP).rejects.toThrow(/aborted/)
    const header = await ti.getHeader()
    expect(header.isBigEndian).toEqual(false)
  })

  it('abort loading a bigwig file (plain abortsignal passed in)', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox.bw'),
    })
    const aborter = new HalfAbortController()
    const indexDataP = ti.getHeader(aborter.signal)
    aborter.abort()
    await expect(indexDataP).rejects.toThrow(/aborted/)
    const header = await ti.getHeader()
    expect(header.isBigEndian).toEqual(false)
  })

  it('test a slow loading getfeatures', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox.bw'),
    })
    const feats = await ti.getFeatures('ctgA', 10000, 40000)
    expect(feats.length).toEqual(30001)
  })

  it('should load a lot of summary data', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/ENCFF826FLP.bw'),
    })
    const feats = await ti.getFeatures('chr22', 0, 51222087, { scale: 0.00001 })
    expect(feats.slice(0, 10)).toMatchSnapshot()
    expect(feats.slice(-10)).toMatchSnapshot()
    expect(feats.length).toEqual(595)
  })

  it('abort with getFeatures', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox.bw'),
    })
    const aborter = new HalfAbortController()
    const ob = ti.getFeatures('ctgA', 0, 100, { signal: aborter.signal })
    aborter.abort()
    await expect(ob).rejects.toThrow(/aborted/)
  })
  it('abort with getFeatureStream', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/volvox.bw'),
    })
    const aborter = new HalfAbortController()
    const ob = await ti.getFeatureStream('ctgA', 0, 100, { signal: aborter.signal })
    aborter.abort()
    await expect(ob.toPromise()).rejects.toThrow(/aborted/)
  })
  it('test uncompressed bw (-unc from wigToBigWig)', async () => {
    const ti = new BigWig({
      path: require.resolve('./data/uncompressed.bw'),
    })
    const ob = await ti.getFeatures('ctgA', 40000, 40100)
    expect(ob.slice(0, 10)).toMatchSnapshot()
  })
})
