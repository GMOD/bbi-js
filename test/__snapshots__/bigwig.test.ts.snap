// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bigwig formats inside file deeply 1`] = `
Array [
  Object {
    "end": 20010,
    "score": 18,
    "start": 20009,
  },
  Object {
    "end": 20011,
    "score": 19,
    "start": 20010,
  },
  Object {
    "end": 20012,
    "score": 18,
    "start": 20011,
  },
  Object {
    "end": 20013,
    "score": 17,
    "start": 20012,
  },
  Object {
    "end": 20014,
    "score": 17,
    "start": 20013,
  },
  Object {
    "end": 20015,
    "score": 16,
    "start": 20014,
  },
  Object {
    "end": 20016,
    "score": 16,
    "start": 20015,
  },
  Object {
    "end": 20017,
    "score": 18,
    "start": 20016,
  },
  Object {
    "end": 20018,
    "score": 17,
    "start": 20017,
  },
  Object {
    "end": 20019,
    "score": 17,
    "start": 20018,
  },
]
`;

exports[`bigwig formats loads a larger bigwig file at different scales 1`] = `
Array [
  Object {
    "end": 2068352,
    "maxScore": 1,
    "minScore": 0,
    "score": 0.004236222847948511,
    "start": 1988800,
    "summary": true,
  },
  Object {
    "end": 2147904,
    "maxScore": 9,
    "minScore": 0,
    "score": 0.12218423169750603,
    "start": 2068352,
    "summary": true,
  },
]
`;

exports[`bigwig formats loads a larger bigwig file at different scales 2`] = `
Array [
  Object {
    "end": 2073477,
    "score": 0,
    "start": 2069069,
  },
  Object {
    "end": 2073485,
    "score": 1,
    "start": 2073477,
  },
  Object {
    "end": 2073564,
    "score": 2,
    "start": 2073485,
  },
  Object {
    "end": 2076947,
    "score": 0,
    "start": 2073564,
  },
  Object {
    "end": 2076956,
    "score": 1,
    "start": 2076947,
  },
  Object {
    "end": 2077030,
    "score": 2,
    "start": 2076956,
  },
  Object {
    "end": 2077034,
    "score": 1,
    "start": 2077030,
  },
  Object {
    "end": 2077775,
    "score": 0,
    "start": 2077034,
  },
  Object {
    "end": 2077844,
    "score": 1,
    "start": 2077775,
  },
  Object {
    "end": 2080490,
    "score": 0,
    "start": 2077844,
  },
]
`;

exports[`bigwig formats loads a larger bigwig file at different scales 3`] = `
Array [
  Object {
    "end": 2073477,
    "score": 0,
    "start": 2069069,
  },
  Object {
    "end": 2073485,
    "score": 1,
    "start": 2073477,
  },
  Object {
    "end": 2073564,
    "score": 2,
    "start": 2073485,
  },
  Object {
    "end": 2076947,
    "score": 0,
    "start": 2073564,
  },
  Object {
    "end": 2076956,
    "score": 1,
    "start": 2076947,
  },
  Object {
    "end": 2077030,
    "score": 2,
    "start": 2076956,
  },
  Object {
    "end": 2077034,
    "score": 1,
    "start": 2077030,
  },
  Object {
    "end": 2077775,
    "score": 0,
    "start": 2077034,
  },
  Object {
    "end": 2077844,
    "score": 1,
    "start": 2077775,
  },
  Object {
    "end": 2080490,
    "score": 0,
    "start": 2077844,
  },
]
`;

exports[`bigwig formats loads a larger bigwig file at different scales 4`] = `
Array [
  Object {
    "end": 2073477,
    "score": 0,
    "start": 2069069,
  },
  Object {
    "end": 2073485,
    "score": 1,
    "start": 2073477,
  },
  Object {
    "end": 2073564,
    "score": 2,
    "start": 2073485,
  },
  Object {
    "end": 2076947,
    "score": 0,
    "start": 2073564,
  },
  Object {
    "end": 2076956,
    "score": 1,
    "start": 2076947,
  },
  Object {
    "end": 2077030,
    "score": 2,
    "start": 2076956,
  },
  Object {
    "end": 2077034,
    "score": 1,
    "start": 2077030,
  },
  Object {
    "end": 2077775,
    "score": 0,
    "start": 2077034,
  },
  Object {
    "end": 2077844,
    "score": 1,
    "start": 2077775,
  },
  Object {
    "end": 2080490,
    "score": 0,
    "start": 2077844,
  },
]
`;

exports[`bigwig formats loads a larger bigwig file at different scales 5`] = `
Array [
  Object {
    "end": 2073477,
    "score": 0,
    "start": 2069069,
  },
  Object {
    "end": 2073485,
    "score": 1,
    "start": 2073477,
  },
  Object {
    "end": 2073564,
    "score": 2,
    "start": 2073485,
  },
  Object {
    "end": 2076947,
    "score": 0,
    "start": 2073564,
  },
  Object {
    "end": 2076956,
    "score": 1,
    "start": 2076947,
  },
  Object {
    "end": 2077030,
    "score": 2,
    "start": 2076956,
  },
  Object {
    "end": 2077034,
    "score": 1,
    "start": 2077030,
  },
  Object {
    "end": 2077775,
    "score": 0,
    "start": 2077034,
  },
  Object {
    "end": 2077844,
    "score": 1,
    "start": 2077775,
  },
  Object {
    "end": 2080490,
    "score": 0,
    "start": 2077844,
  },
]
`;

exports[`bigwig formats loads bedgraph bigwig file 1`] = `
Array [
  Object {
    "end": 3,
    "score": 1,
    "start": 2,
  },
  Object {
    "end": 4,
    "score": 1,
    "start": 3,
  },
  Object {
    "end": 5,
    "score": 1,
    "start": 4,
  },
  Object {
    "end": 6,
    "score": 1,
    "start": 5,
  },
  Object {
    "end": 7,
    "score": 1,
    "start": 6,
  },
  Object {
    "end": 8,
    "score": 2,
    "start": 7,
  },
  Object {
    "end": 9,
    "score": 2,
    "start": 8,
  },
  Object {
    "end": 10,
    "score": 2,
    "start": 9,
  },
  Object {
    "end": 11,
    "score": 2,
    "start": 10,
  },
  Object {
    "end": 12,
    "score": 2,
    "start": 11,
  },
  Object {
    "end": 13,
    "score": 2,
    "start": 12,
  },
  Object {
    "end": 14,
    "score": 2,
    "start": 13,
  },
  Object {
    "end": 15,
    "score": 2,
    "start": 14,
  },
  Object {
    "end": 16,
    "score": 3,
    "start": 15,
  },
  Object {
    "end": 17,
    "score": 4,
    "start": 16,
  },
  Object {
    "end": 18,
    "score": 5,
    "start": 17,
  },
  Object {
    "end": 19,
    "score": 5,
    "start": 18,
  },
  Object {
    "end": 20,
    "score": 5,
    "start": 19,
  },
  Object {
    "end": 21,
    "score": 5,
    "start": 20,
  },
  Object {
    "end": 22,
    "score": 5,
    "start": 21,
  },
  Object {
    "end": 23,
    "score": 5,
    "start": 22,
  },
  Object {
    "end": 24,
    "score": 5,
    "start": 23,
  },
  Object {
    "end": 25,
    "score": 5,
    "start": 24,
  },
  Object {
    "end": 26,
    "score": 5,
    "start": 25,
  },
  Object {
    "end": 27,
    "score": 5,
    "start": 26,
  },
  Object {
    "end": 28,
    "score": 5,
    "start": 27,
  },
  Object {
    "end": 29,
    "score": 5,
    "start": 28,
  },
  Object {
    "end": 30,
    "score": 5,
    "start": 29,
  },
  Object {
    "end": 31,
    "score": 5,
    "start": 30,
  },
  Object {
    "end": 32,
    "score": 5,
    "start": 31,
  },
  Object {
    "end": 33,
    "score": 5,
    "start": 32,
  },
  Object {
    "end": 34,
    "score": 5,
    "start": 33,
  },
  Object {
    "end": 35,
    "score": 5,
    "start": 34,
  },
  Object {
    "end": 36,
    "score": 5,
    "start": 35,
  },
  Object {
    "end": 37,
    "score": 5,
    "start": 36,
  },
  Object {
    "end": 38,
    "score": 5,
    "start": 37,
  },
  Object {
    "end": 39,
    "score": 5,
    "start": 38,
  },
  Object {
    "end": 40,
    "score": 5,
    "start": 39,
  },
  Object {
    "end": 41,
    "score": 5,
    "start": 40,
  },
  Object {
    "end": 42,
    "score": 5,
    "start": 41,
  },
  Object {
    "end": 43,
    "score": 6,
    "start": 42,
  },
  Object {
    "end": 44,
    "score": 7,
    "start": 43,
  },
  Object {
    "end": 45,
    "score": 7,
    "start": 44,
  },
  Object {
    "end": 46,
    "score": 7,
    "start": 45,
  },
  Object {
    "end": 47,
    "score": 7,
    "start": 46,
  },
  Object {
    "end": 48,
    "score": 8,
    "start": 47,
  },
  Object {
    "end": 49,
    "score": 8,
    "start": 48,
  },
  Object {
    "end": 50,
    "score": 8,
    "start": 49,
  },
  Object {
    "end": 51,
    "score": 8,
    "start": 50,
  },
  Object {
    "end": 52,
    "score": 8,
    "start": 51,
  },
  Object {
    "end": 53,
    "score": 8,
    "start": 52,
  },
  Object {
    "end": 54,
    "score": 8,
    "start": 53,
  },
  Object {
    "end": 55,
    "score": 8,
    "start": 54,
  },
  Object {
    "end": 56,
    "score": 9,
    "start": 55,
  },
  Object {
    "end": 57,
    "score": 9,
    "start": 56,
  },
  Object {
    "end": 58,
    "score": 9,
    "start": 57,
  },
  Object {
    "end": 59,
    "score": 9,
    "start": 58,
  },
  Object {
    "end": 60,
    "score": 9,
    "start": 59,
  },
  Object {
    "end": 61,
    "score": 9,
    "start": 60,
  },
  Object {
    "end": 62,
    "score": 9,
    "start": 61,
  },
  Object {
    "end": 63,
    "score": 9,
    "start": 62,
  },
  Object {
    "end": 64,
    "score": 9,
    "start": 63,
  },
  Object {
    "end": 65,
    "score": 9,
    "start": 64,
  },
  Object {
    "end": 66,
    "score": 10,
    "start": 65,
  },
  Object {
    "end": 67,
    "score": 10,
    "start": 66,
  },
  Object {
    "end": 68,
    "score": 10,
    "start": 67,
  },
  Object {
    "end": 69,
    "score": 10,
    "start": 68,
  },
  Object {
    "end": 70,
    "score": 10,
    "start": 69,
  },
  Object {
    "end": 71,
    "score": 10,
    "start": 70,
  },
  Object {
    "end": 72,
    "score": 10,
    "start": 71,
  },
  Object {
    "end": 73,
    "score": 10,
    "start": 72,
  },
  Object {
    "end": 74,
    "score": 10,
    "start": 73,
  },
  Object {
    "end": 75,
    "score": 10,
    "start": 74,
  },
  Object {
    "end": 76,
    "score": 10,
    "start": 75,
  },
  Object {
    "end": 77,
    "score": 11,
    "start": 76,
  },
  Object {
    "end": 78,
    "score": 11,
    "start": 77,
  },
  Object {
    "end": 79,
    "score": 11,
    "start": 78,
  },
  Object {
    "end": 80,
    "score": 11,
    "start": 79,
  },
  Object {
    "end": 81,
    "score": 11,
    "start": 80,
  },
  Object {
    "end": 82,
    "score": 11,
    "start": 81,
  },
  Object {
    "end": 83,
    "score": 11,
    "start": 82,
  },
  Object {
    "end": 84,
    "score": 11,
    "start": 83,
  },
  Object {
    "end": 85,
    "score": 11,
    "start": 84,
  },
  Object {
    "end": 86,
    "score": 11,
    "start": 85,
  },
  Object {
    "end": 87,
    "score": 11,
    "start": 86,
  },
  Object {
    "end": 88,
    "score": 11,
    "start": 87,
  },
  Object {
    "end": 89,
    "score": 11,
    "start": 88,
  },
  Object {
    "end": 90,
    "score": 11,
    "start": 89,
  },
  Object {
    "end": 91,
    "score": 12,
    "start": 90,
  },
  Object {
    "end": 92,
    "score": 12,
    "start": 91,
  },
  Object {
    "end": 93,
    "score": 12,
    "start": 92,
  },
  Object {
    "end": 94,
    "score": 12,
    "start": 93,
  },
  Object {
    "end": 95,
    "score": 13,
    "start": 94,
  },
  Object {
    "end": 96,
    "score": 13,
    "start": 95,
  },
  Object {
    "end": 97,
    "score": 13,
    "start": 96,
  },
  Object {
    "end": 98,
    "score": 13,
    "start": 97,
  },
  Object {
    "end": 99,
    "score": 13,
    "start": 98,
  },
  Object {
    "end": 100,
    "score": 13,
    "start": 99,
  },
]
`;

exports[`bigwig formats loads bedgraph bigwig file 2`] = `
Array [
  Object {
    "end": 162,
    "maxScore": 13,
    "minScore": 1,
    "score": 9.0125,
    "start": 2,
    "summary": true,
  },
]
`;

exports[`bigwig formats loads bedgraph bigwig file 3`] = `
Array [
  Object {
    "end": 642,
    "maxScore": 24,
    "minScore": 1,
    "score": 9.746875,
    "start": 2,
    "summary": true,
  },
]
`;

exports[`bigwig formats loads bedgraph bigwig file 4`] = `
Array [
  Object {
    "end": 2562,
    "maxScore": 32,
    "minScore": 13,
    "score": 21.7625,
    "start": 1922,
    "summary": true,
  },
]
`;

exports[`bigwig formats loads fixedstep bigwig 1`] = `
Array [
  Object {
    "end": 100,
    "score": 281,
    "start": 0,
  },
  Object {
    "end": 200,
    "score": 183,
    "start": 100,
  },
  Object {
    "end": 300,
    "score": 213,
    "start": 200,
  },
  Object {
    "end": 400,
    "score": 191,
    "start": 300,
  },
  Object {
    "end": 500,
    "score": 288,
    "start": 400,
  },
  Object {
    "end": 600,
    "score": 184,
    "start": 500,
  },
  Object {
    "end": 700,
    "score": 258,
    "start": 600,
  },
  Object {
    "end": 800,
    "score": 173,
    "start": 700,
  },
  Object {
    "end": 900,
    "score": 274,
    "start": 800,
  },
  Object {
    "end": 1000,
    "score": 292,
    "start": 900,
  },
]
`;

exports[`bigwig formats loads fixedstep bigwig 2`] = `
Array [
  Object {
    "end": 100,
    "score": 281,
    "start": 0,
  },
  Object {
    "end": 200,
    "score": 183,
    "start": 100,
  },
  Object {
    "end": 300,
    "score": 213,
    "start": 200,
  },
  Object {
    "end": 400,
    "score": 191,
    "start": 300,
  },
  Object {
    "end": 500,
    "score": 288,
    "start": 400,
  },
  Object {
    "end": 600,
    "score": 184,
    "start": 500,
  },
  Object {
    "end": 700,
    "score": 258,
    "start": 600,
  },
  Object {
    "end": 800,
    "score": 173,
    "start": 700,
  },
  Object {
    "end": 900,
    "score": 274,
    "start": 800,
  },
  Object {
    "end": 1000,
    "score": 292,
    "start": 900,
  },
]
`;

exports[`bigwig formats loads fixedstep bigwig 3`] = `
Array [
  Object {
    "end": 100,
    "score": 281,
    "start": 0,
  },
  Object {
    "end": 200,
    "score": 183,
    "start": 100,
  },
  Object {
    "end": 300,
    "score": 213,
    "start": 200,
  },
  Object {
    "end": 400,
    "score": 191,
    "start": 300,
  },
  Object {
    "end": 500,
    "score": 288,
    "start": 400,
  },
  Object {
    "end": 600,
    "score": 184,
    "start": 500,
  },
  Object {
    "end": 700,
    "score": 258,
    "start": 600,
  },
  Object {
    "end": 800,
    "score": 173,
    "start": 700,
  },
  Object {
    "end": 900,
    "score": 274,
    "start": 800,
  },
  Object {
    "end": 1000,
    "score": 292,
    "start": 900,
  },
]
`;

exports[`bigwig formats loads fixedstep bigwig 4`] = `
Array [
  Object {
    "end": 2000,
    "score": 192,
    "start": 1900,
  },
  Object {
    "end": 2100,
    "score": 114,
    "start": 2000,
  },
]
`;

exports[`bigwig formats loads variable step bigwig 1`] = `
Array [
  Object {
    "end": 12,
    "score": 0.7819849848747253,
    "start": 11,
  },
  Object {
    "end": 16,
    "score": 0.8109930157661438,
    "start": 15,
  },
  Object {
    "end": 20,
    "score": 0.7696009874343872,
    "start": 19,
  },
  Object {
    "end": 24,
    "score": 0.733830988407135,
    "start": 23,
  },
  Object {
    "end": 28,
    "score": 0.7320680022239685,
    "start": 27,
  },
  Object {
    "end": 32,
    "score": 0.5884150266647339,
    "start": 31,
  },
  Object {
    "end": 36,
    "score": 0.5306519865989685,
    "start": 35,
  },
  Object {
    "end": 40,
    "score": 0.36328500509262085,
    "start": 39,
  },
  Object {
    "end": 44,
    "score": 0.1721539944410324,
    "start": 43,
  },
  Object {
    "end": 48,
    "score": 0.06392999738454819,
    "start": 47,
  },
]
`;

exports[`bigwig formats missing data 1`] = `
Array [
  Object {
    "end": 4210,
    "score": 20,
    "start": 4209,
  },
  Object {
    "end": 4211,
    "score": 20,
    "start": 4210,
  },
  Object {
    "end": 4212,
    "score": 20,
    "start": 4211,
  },
  Object {
    "end": 4213,
    "score": 21,
    "start": 4212,
  },
  Object {
    "end": 4214,
    "score": 21,
    "start": 4213,
  },
  Object {
    "end": 4215,
    "score": 21,
    "start": 4214,
  },
  Object {
    "end": 4216,
    "score": 22,
    "start": 4215,
  },
  Object {
    "end": 4217,
    "score": 22,
    "start": 4216,
  },
  Object {
    "end": 4218,
    "score": 22,
    "start": 4217,
  },
  Object {
    "end": 4219,
    "score": 21,
    "start": 4218,
  },
]
`;

exports[`bigwig formats missing data 2`] = `
Array [
  Object {
    "end": 5200,
    "score": 19,
    "start": 5199,
  },
  Object {
    "end": 5201,
    "score": 19,
    "start": 5200,
  },
  Object {
    "end": 5202,
    "score": 19,
    "start": 5201,
  },
  Object {
    "end": 5203,
    "score": 20,
    "start": 5202,
  },
  Object {
    "end": 5204,
    "score": 19,
    "start": 5203,
  },
  Object {
    "end": 5205,
    "score": 18,
    "start": 5204,
  },
  Object {
    "end": 5206,
    "score": 18,
    "start": 5205,
  },
  Object {
    "end": 5207,
    "score": 19,
    "start": 5206,
  },
  Object {
    "end": 5208,
    "score": 18,
    "start": 5207,
  },
  Object {
    "end": 5209,
    "score": 18,
    "start": 5208,
  },
]
`;

exports[`bigwig formats should load a lot of summary data 1`] = `
Array [
  Object {
    "end": 86144,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 0,
    "summary": true,
  },
  Object {
    "end": 172288,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 86144,
    "summary": true,
  },
  Object {
    "end": 258432,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 172288,
    "summary": true,
  },
  Object {
    "end": 344576,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 258432,
    "summary": true,
  },
  Object {
    "end": 430720,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 344576,
    "summary": true,
  },
  Object {
    "end": 516864,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 430720,
    "summary": true,
  },
  Object {
    "end": 603008,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 516864,
    "summary": true,
  },
  Object {
    "end": 689152,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 603008,
    "summary": true,
  },
  Object {
    "end": 775296,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 689152,
    "summary": true,
  },
  Object {
    "end": 861440,
    "maxScore": 0,
    "minScore": 0,
    "score": 0,
    "start": 775296,
    "summary": true,
  },
]
`;

exports[`bigwig formats should load a lot of summary data 2`] = `
Array [
  Object {
    "end": 50480384,
    "maxScore": 19.640100479125977,
    "minScore": 0,
    "score": 0.09248444591122575,
    "start": 50394240,
    "summary": true,
  },
  Object {
    "end": 50566528,
    "maxScore": 29.029600143432617,
    "minScore": 0,
    "score": 0.3248449502196365,
    "start": 50480384,
    "summary": true,
  },
  Object {
    "end": 50652672,
    "maxScore": 44.189998626708984,
    "minScore": 0,
    "score": 1.4510407430813634,
    "start": 50566528,
    "summary": true,
  },
  Object {
    "end": 50738816,
    "maxScore": 71.71600341796875,
    "minScore": 0,
    "score": 2.0017577412787166,
    "start": 50652672,
    "summary": true,
  },
  Object {
    "end": 50824960,
    "maxScore": 68.74019622802734,
    "minScore": 0,
    "score": 1.156110973070188,
    "start": 50738816,
    "summary": true,
  },
  Object {
    "end": 50911104,
    "maxScore": 34.37009811401367,
    "minScore": 0,
    "score": 1.481282179458206,
    "start": 50824960,
    "summary": true,
  },
  Object {
    "end": 50997248,
    "maxScore": 70.96739959716797,
    "minScore": 0,
    "score": 1.4128606966790562,
    "start": 50911104,
    "summary": true,
  },
  Object {
    "end": 51083392,
    "maxScore": 41.391300201416016,
    "minScore": 0,
    "score": 0.7212576265548543,
    "start": 50997248,
    "summary": true,
  },
  Object {
    "end": 51169536,
    "maxScore": 9.819890022277832,
    "minScore": 0,
    "score": 0.20769502693705816,
    "start": 51083392,
    "summary": true,
  },
  Object {
    "end": 51255680,
    "maxScore": 24.54990005493164,
    "minScore": 0,
    "score": 0.0633376828883356,
    "start": 51169536,
    "summary": true,
  },
]
`;

exports[`bigwig formats test uncompressed bw (-unc from wigToBigWig) 1`] = `
Array [
  Object {
    "end": 40000,
    "score": 23,
    "start": 39999,
  },
  Object {
    "end": 40001,
    "score": 23,
    "start": 40000,
  },
  Object {
    "end": 40002,
    "score": 23,
    "start": 40001,
  },
  Object {
    "end": 40003,
    "score": 23,
    "start": 40002,
  },
  Object {
    "end": 40004,
    "score": 25,
    "start": 40003,
  },
  Object {
    "end": 40005,
    "score": 25,
    "start": 40004,
  },
  Object {
    "end": 40006,
    "score": 24,
    "start": 40005,
  },
  Object {
    "end": 40007,
    "score": 23,
    "start": 40006,
  },
  Object {
    "end": 40008,
    "score": 22,
    "start": 40007,
  },
  Object {
    "end": 40009,
    "score": 22,
    "start": 40008,
  },
]
`;
